README.md
**[BUILD]**

```shell
First make please use sudo 
sudo make -j
```

**[RUN GMSL CAMREA DEMO]**

使用NV自带接口时，参考cameras_egl_demo
使用miivii提供的接口读取图像以及当前图像时间戳时，参考cameras_sdk_demo
使用opencv接口获取相机图像，同时调用miivii接口获取图像时间戳时，参考cameras_opencv_demo
使用miivii接口获取图像以及当前图像时间戳，并不进行opencv库调用时，参考cameras_sdk_noopencv_demo

```shell
./bin/cameras_opencv_demo -s 1280x720 -d /dev/video0 -n 4-4 -r 30-30 -b 0xf-0xf0 -p 0-0-0-0-1-2-3-4
./bin/cameras_sdk_demo -s 1280x720 -d /dev/video0 -n 4-4 -r 30-30 -b 0xf-0xf0 -p 0-0-0-0-1-2-3-4
./bin/cameras_egl_demo -d /dev/video0 -s 1280x720
```
parameter comment:
-d:Set output resolution of video device
-n:Set sync and async camera no. for example: [-n 2-4] the forward one is sync cameras no 2,the after one is async cameras no 4 (8 sync cameras is setted by default.like [-n 8-0])
-r:Set sync and async camera freq for example: [-r 30-20] the forward one is sync cameras freq 30,the after one is async cameras freq 20(sync freq 30 is setted by default.like [-r 30-0])
-b:Set which cameras you want to trigger.example: [-b 0x0f-0xf0] the forward one is sync cameras which you want trigger,the after one is async cameras which you want trigger(all 8 cameras is setted sync model by default.like[-b 0xff-0])
-p:Set async cameras is triggered at which angle in a circle,not set by default.
-h:show help.

同步相机开启方式（以三个摄像头为例）：
./bin/cameras_sdk_demo -d /dev/video0 -s 1280x720 -m 3 -r 30-0
-d：设置打开的第一个摄像头设备号。
-s：设置采集摄像头的分辨率，中间以小写x分隔。
-m：设置打开的摄像头个数。
-r：设置外部触发摄像头的采集频率，“-”前为同步频率，“-”后为异步频率。此例中只打开同步摄像头，所以只设置“-”前的频率为30。


**[SDK CAPACITY]**

GMSL-SDK support custom output image resolutions as well 
 as multiple formats.
 - UYVY
 - YUYV
 - YVYU
 - YUV420M
 - XBGR32
 - ARGB32

 The image format is set to ABGR32 by default.

 Notice that：
 > YUV color system images should only be obtained by method MVGmslGetImage() while RGB color 
 system images should only be obtained by method MVGmslGetCvMat();


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [Unreleased]
## [1.0.0] - 2020-07-14
- Adapted for Jetpack 4.4 GA
- refactor for better performance 

## [0.1.4] - 2019-08-07
- Fixed bugs in cameras_opencv_demo.cpp when set fmt as YUYV results program crash.
- Added version number support in both dynamic libs and header file.
- Removed redundant outputs.

## [0.1.3] - 2019-06-18
- Change the directory structure for easy management.

## [0.1.2] - 2019-04-13
- Adjust folder structure to adapt C++ Project Standard.
- Optimize user compilation methods
  
## [0.1.1] - 2019-04-09

### Changed

- Added timestamp interface in SDK and usage in example.cpp.
- Modified make logic. 

## [0.1.0] - 2019-03-08

### Changed

- Rewrote the SDK so that it no longer depends on Nv header files.
- Modified example.cpp.
- Modified folder structure.

## [0.0.5] - 2019-02-20
### Changed
- Add example launch command in README.md.

## [0.0.4] - 2019-02-16
### Changed
- S2-pro compatible.

## [0.0.3] - 2019-01-17
### Added
- R5 synchronize support

### Changed
- Use CV Display instead of EGLRender.
### Fixed
- Camera frozen in long time test.

## [0.0.2] - 2018-12-20

### Removed
- Removed SDK alone with example.

## [0.0.1] - 2018-12-18
### Added
- Added control of multi-window parameters. 
- Fix a build problem on build.sh script. 
### Changed
- Make install path changed to /lib/
- Simplify code structure.
- Replace multi-media framework to xavier compatible.


samples/cameras_egl_demo/camera_v4l2_cuda.h
/*
 * Copyright (c) 2016,2017, NVIDIA CORPORATION. All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *  * Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 *  * Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *  * Neither the name of NVIDIA CORPORATION nor the names of its
 *    contributors may be used to endorse or promote products derived
 *    from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
 * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
 * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
 * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
 * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <queue>
#include "NvJpegDecoder.h"

#define V4L2_BUFFERS_NUM    4

#define INFO(fmt, ...) \
    if (ctx->enable_verbose == true) \
        printf("INFO: " fmt "\n", ##__VA_ARGS__);
//        printf("INFO: %s(): (line:%d) " fmt "\n", __FUNCTION__, __LINE__, ##__VA_ARGS__);

#define WARN(fmt, ...) \
    if (ctx->enable_verbose == true) \
        printf("WARN: " fmt "\n", ##__VA_ARGS__);
//        printf("WARN: %s(): (line:%d) " fmt "\n", __FUNCTION__, __LINE__, ##__VA_ARGS__);

#define CHECK_ERROR(cond, label, fmt, ...) \
    if (!cond) { \
        error = 1; \
        printf("ERROR: %s(): (line:%d) " fmt "\n", __FUNCTION__, __LINE__, ##__VA_ARGS__); \
        goto label; \
    }

#define ERROR_RETURN(fmt, ...) \
    do { \
        printf("ERROR: %s(): (line:%d) " fmt "\n", __FUNCTION__, __LINE__, ##__VA_ARGS__); \
        return false; \
    } while(0)

typedef struct
{
    // Hold the user accessible pointer
    unsigned char * start;
    // Hold the memory length
    unsigned int size;
    // Hold the file descriptor of NvBuffer
    int dmabuff_fd;
} nv_buffer;

typedef struct
{
    // camera v4l2 context
    const char * cam_devname;
    char cam_file[16];
    int cam_fd;
    unsigned int cam_pixfmt;
    unsigned int cam_w;
    unsigned int cam_h;
    unsigned int win_w;
    unsigned int win_h;
    unsigned int win_t;
    unsigned int win_l;
    unsigned int frame;
    unsigned int save_n_frame;

    // Global buffer ptr
    nv_buffer * g_buff;
    bool capture_dmabuf;

    // EGL renderer
    NvEglRenderer *renderer;
    int render_dmabuf_fd;
    int fps;

    // CUDA processing
    bool enable_cuda;
    EGLDisplay egl_display;
    EGLImageKHR egl_image;

    // MJPEG decoding
    NvJPEGDecoder *jpegdec;

    // Verbose option
    bool enable_verbose;

} context_t;

// Correlate v4l2 pixel format and NvBuffer color format
typedef struct
{
    unsigned int v4l2_pixfmt;
    NvBufferColorFormat nvbuff_color;
} nv_color_fmt;


samples/cameras_opencv_demo/cameras_opencv_demo.cpp
#include <opencv2/core.hpp>
#include <opencv2/videoio.hpp>
#include <opencv2/highgui.hpp>
#include <opencv2/imgproc.hpp>  // cv::Canny()
#include <iostream>
#include <csignal>
#include <fstream>
#include <zconf.h>
#include <string>
#include <chrono>

#include "MvGmslCamera.h"

using namespace cv;using std::string;
using std::cout; using std::cerr; using std::endl;



struct sample_context{
    string cam_devname;
    string pix_fmt;
    unsigned int cam_w{1280};
    unsigned int cam_h{720};
    unsigned char camera_no;
    struct sync_out_a_cfg_client_t stCameraCfgSend{};
};

static void
print_usage(void) {
    printf("\n\tUsage: example [OPTIONS]\n\n"
           "\tExample: \n"
           "\t./cameras_opencv_demo -d /dev/video0 -s 1280x720\n\n"
           "\tSupported options:\n"
           "\t-d\t\tSet V4l2 video device node\n"
           "\t-s\t\tSet output resolution of video device\n"
           "\t-n\t\tSet sync and async camera no. for example: [-n 2-4] the forward one is sync cameras no 2,the after one is async cameras no 4 (8 sync cameras is setted by default.like [-n 8-0])\n"
           "\t-r\t\tSet sync and async camera freq for example: [-r 30-20] the forward one is sync cameras freq 30,the after one is async cameras freq 20(sync freq 30 is setted by default.like [-r 30-0]) \n"
           "\t-b\t\tSet which cameras you want to trigger.example: [-b 0x0f-0xf0] the forward one is sync cameras which you want trigger,the after one is async cameras which you want trigger(all 8 cameras is setted sync model by default.like[-b 0xff-0])\n"
           "\t-p\t\tSet async cameras is triggered at which angle in a circle,not set by default.\n"
           "\t-h\t\tPrint this usage\n\n"
           "\tNOTE: It runs infinitely until you terminate it with <ctrl+c>\n");
}

static bool
parse_cmdline(struct sample_context *ctx, int argc, char **argv) {
    int c;
    int sync_camera_bit_draw = 0,async_camera_bit_draw = 0;

    if (argc < 2) {
        print_usage();
        exit(EXIT_SUCCESS);
    }

    while ((c = getopt(argc, argv, "d:s:r:n:b:f:p:h")) != -1) {
        switch (c) {
            case 'd':
                ctx->cam_devname = optarg;
        sscanf(ctx->cam_devname.c_str(),"/dev/video%hhu",&ctx->camera_no);
                break;
            case 's':
                if (sscanf(optarg, "%ux%u",&ctx->cam_w, &ctx->cam_h) != 2) {
                    print_usage();
                    return false;
                }
                break;
            case 'r':
                if (sscanf(optarg, "%hhu-%hhu",&ctx->stCameraCfgSend.sync_freq, &ctx->stCameraCfgSend.async_freq) != 2) {
                   print_usage();
                    return false;
                }
                           printf("sync_freq : %d async_freq:%d\n",ctx->stCameraCfgSend.sync_freq,ctx->stCameraCfgSend.async_freq);
                break;
            case 'n':
                if (sscanf(optarg, "%hhu-%hhu",&ctx->stCameraCfgSend.sync_camera_num, &ctx->stCameraCfgSend.async_camera_num) != 2) {
                   print_usage();
                    return false;
                }
                           printf("sync_camera_num : %d async_camera_num:%d\n",(ctx->stCameraCfgSend.sync_camera_num),(ctx->stCameraCfgSend.async_camera_num));
                break;
            case 'b':
                if (sscanf(optarg, "%x-%x",&sync_camera_bit_draw, &async_camera_bit_draw) != 2) {
                   print_usage();
                   return false;
                }
                   ctx->stCameraCfgSend.sync_camera_bit_draw = (unsigned char)sync_camera_bit_draw;
                   ctx->stCameraCfgSend.async_camera_bit_draw = (unsigned char)async_camera_bit_draw;
                   printf("sync_camera_bit_draw : %d async_camera_bit_draw:%d\n",ctx->stCameraCfgSend.sync_camera_bit_draw,ctx->stCameraCfgSend.async_camera_bit_draw);
                break;
           case 'p':
               if (sscanf(optarg, "%hhu-%hhu-%hhu-%hhu-%hhu-%hhu-%hhu-%hhu",
                          &ctx->stCameraCfgSend.async_camera_pos[0], &ctx->stCameraCfgSend.async_camera_pos[1],&ctx->stCameraCfgSend.async_camera_pos[2],
                          &ctx->stCameraCfgSend.async_camera_pos[3],&ctx->stCameraCfgSend.async_camera_pos[4],&ctx->stCameraCfgSend.async_camera_pos[5]
                          ,&ctx->stCameraCfgSend.async_camera_pos[6],&ctx->stCameraCfgSend.async_camera_pos[7]) != 8) {
                   print_usage();
                   return false;
               }
                  printf("pos:[0]:%hhu [1]:%hhu [2]:%hhu [3]:%hhu [4]:%hhu [5]:%hhu [6]:%hhu [7]:%hhu \n",ctx->stCameraCfgSend.async_camera_pos[0],ctx->stCameraCfgSend.async_camera_pos[1],ctx->stCameraCfgSend.async_camera_pos[2],
                  ctx->stCameraCfgSend.async_camera_pos[3],ctx->stCameraCfgSend.async_camera_pos[4],ctx->stCameraCfgSend.async_camera_pos[5],ctx->stCameraCfgSend.async_camera_pos[6],ctx->stCameraCfgSend.async_camera_pos[7]);
               break;
            case 'f':
                ctx->pix_fmt = optarg;
                break;
            case 'h':
                print_usage();
                exit(EXIT_SUCCESS);
                break;
            default:
                print_usage();
                return false;
        }
    }
    return true;
}

int main(int argc, char *argv[])
{
    struct sample_context ctx = {};
    ctx.stCameraCfgSend.async_camera_num = 0;
    ctx.stCameraCfgSend.async_freq = 0;
    ctx.stCameraCfgSend.async_camera_bit_draw = 0;
    ctx.stCameraCfgSend.sync_camera_num = 8;
    ctx.stCameraCfgSend.sync_freq = 30;
    ctx.stCameraCfgSend.sync_camera_bit_draw = 0xff;
    ctx.cam_devname = "/dev/video0";
    ctx.cam_w = 1280;
    ctx.cam_h = 720;
    ctx.pix_fmt = "UYVY";
    
    if (!parse_cmdline(&ctx, argc, argv)) {
            return -1;
    }

    char Gstring[256];
    sprintf(Gstring,"v4l2src device=%s ! video/x-raw, width=%d, height=%d, format=UYVY ! appsink",ctx.cam_devname.c_str(),ctx.cam_w,ctx.cam_h);
    miivii::MvGmslCamera mvcam(ctx.stCameraCfgSend);
    cout << "Opening camera..." << endl;
    VideoCapture capture(Gstring, CAP_GSTREAMER);
    if (!capture.isOpened())
    {
        cerr << "ERROR: Can't initialize camera capture" << endl;
        return 1;
    }
    Mat frame;
    Mat frameRaw;
    bool stop = false;
    uint64_t timestamp = 0;
    unsigned int colorConvert;
    if(ctx.pix_fmt == "YUYV")
    {
        colorConvert = COLOR_YUV2BGR_YUYV;
    }
    else
    {
        colorConvert = COLOR_YUV2BGR_UYVY;
    }

    while(!stop)
    {
        mvcam.GetGmslTimeStamp(ctx.camera_no,timestamp);
        capture.read(frameRaw);
        cvtColor(frameRaw,frame,colorConvert);
        namedWindow("Video" + std::to_string(ctx.camera_no), 0);
        resizeWindow("Video" + std::to_string(ctx.camera_no), 320, 180);
        imshow("Video" + std::to_string(ctx.camera_no),frame);
        std::cout << "Timestamp from MIIVII  :  " << timestamp << std::endl;
        if(waitKey(1)==27)
        {
            stop = true;
        }
    }
}


samples/cameras_sdk_demo/cameras_sdk_demo.cpp
//
// Created by alex on 18-12-8.
//
//
#include <opencv2/opencv.hpp>
#include <opencv2/core/core.hpp>
#include <opencv2/highgui/highgui.hpp>
#include <string>
#include <iostream>
#include <zconf.h>
#include <csignal>
#include <thread>
#include "MvGmslCamera.h"
#include <fstream>
#include <chrono>

using namespace std::chrono;

using std::string;
sig_atomic_t exitRequested = 0;
uint camera_num = 1;

struct sync_out_a_cfg_client_t stCameraCfgSend = {};

char dev_node[32] = "/dev/video0";/*不输入-d参数时，默认开启的第一个的设备*/
std::string camera_fmt_str = "UYVY";/*默认采集的图像格式*/
std::string output_fmt_str = "ABGR32";/*默认输出的图像格式，支持类型见README*/
uint cam_w = 1280;/*不输入-s参数时。默认采集摄像头的分辨率*/
uint cam_h = 720;/*不输入-s参数时。默认采集摄像头的分辨率*/
uint out_w = 320;/*屏幕输出的视频分辨率*/
uint out_h = 180;/*屏幕输出的视频分辨率*/
uint64_t timestampbefore[8] = {0};/*上一次采集图像时采集时间*/
uint64_t LinuxGetFrameTimeBefore[8] = {0};/*上一次采集图像时的系统时间*/

void handler(int) {
    std::cout << "will exit..." << std::endl;
    exitRequested = true;
}
/*图像采集时的时间戳记录，将时间戳间隔与帧数不相符的相关时间信息打印到/tmp/cameras_sdk_demo.log文件中，每个设备生成单独log文件*/
void CheckTimeStampLog(uint64_t timestamp,uint8_t camera_no)
{
    uint64_t FrameInterval = 0;
    char buffer[256] = {0};
    uint64_t LinuxFrameInterval{};
    struct timeval cur_time;
    uint64_t LinuxGetFrameTime{};
    uint64_t time_interval{};
    uint64_t FrameTransferDelay{};
    FILE * file_diff = NULL;
    char file_name[100] = {0};
    if(0 == timestamp)
    {
        /*camera Data is not available during camera preparation*/
        return;
    }
    FrameInterval = timestamp - timestampbefore[camera_no];
    timestampbefore[camera_no] = timestamp;
    gettimeofday(&cur_time, NULL);
    LinuxGetFrameTime = cur_time.tv_sec * 1000000000 + cur_time.tv_usec * 1000;
    LinuxFrameInterval = LinuxGetFrameTime - LinuxGetFrameTimeBefore[camera_no];
    LinuxGetFrameTimeBefore[camera_no] = LinuxGetFrameTime;
    FrameTransferDelay = LinuxGetFrameTime - timestamp;
    if(stCameraCfgSend.sync_freq != 0)
    time_interval = 1000000000 / stCameraCfgSend.sync_freq;
    else{
        time_interval = 1000000000 / stCameraCfgSend.async_freq;
    }
/*    if((FrameInterval > (time_interval + 15000000) || FrameInterval < (time_interval - 15000000)) 
        || (LinuxFrameInterval > (time_interval + 15000000) || LinuxFrameInterval < (time_interval - 15000000)) 
        || FrameTransferDelay < 70000000 || FrameTransferDelay > 90000000)
    {
        printf("camera_no==========%d\n",camera_no);
        printf("timestamp==========%ld\n",timestamp/1000000);
        printf("FrameInterva===-------------=======%ld\n",FrameInterval/1000000);
        printf("LinuxGetFrameTime======------------====%ld\n",LinuxGetFrameTime/1000000);
        printf("LinuxFrameInterval======---------------====%ld\n",LinuxFrameInterval/1000000);
        printf("FrameTransferDelay======-----------------------------------====%ld\n",FrameTransferDelay/1000000);
    }*/
    if(((FrameInterval > (time_interval + 12000000) || FrameInterval < (time_interval - 12000000))) && (FrameInterval != 0))
    {
        sprintf(file_name,"/tmp/cameras_sdk_demo_video%d.log",camera_no);
        file_diff = fopen(file_name,"a+");
        sprintf(buffer,"Timestamp : %ld FrameInterval  :  %ld FrameTransferDelay : %ld LinuxGetFrameTime : %ld LinuxFrameInterval : %ld\n"
                        ,timestamp,FrameInterval,FrameTransferDelay,LinuxGetFrameTime,LinuxFrameInterval);
        fwrite(buffer,sizeof(char),strlen(buffer),file_diff);
        fflush(file_diff);
        fclose(file_diff);
    }
    if(atoi(getenv("CHECK_TIME")))
    {
        printf("Timestamp : %ld FrameInterval : %ld FrameTransferDelay : %ld   LinuxGetFrameTime : %ld LinuxFrameInterval : %ld\n"
                        ,timestamp,FrameInterval,FrameTransferDelay,LinuxGetFrameTime,LinuxFrameInterval);
    }
}
/*demo程序命令相关参数介绍，详细解释见README*/
static void
print_usage(void) {
    printf("\n\tUsage: example [OPTIONS]\n\n"
           "\tExample: \n"
           "\t./cameras_sdk_demo -d /dev/video0 -s 1280x720\n\n"
           "\tSupported options:\n"
           "\t-d\t\tSet V4l2 video device node\n"
           "\t-m\t\tSet V4l2 video num\n"
           "\t-s\t\tSet output resolution of video device\n"
           "\t-n\t\tSet sync and async camera no. for example: [-n 2-4] the forward one is sync cameras no 2,the after one is async cameras no 4 (8 sync cameras is setted by default.like [-n 8-0])\n"
           "\t-r\t\tSet sync and async camera freq for example: [-r 30-20] the forward one is sync cameras freq 30,the after one is async cameras freq 20(sync freq 30 is setted by default.like [-r 30-0]) \n"
           "\t-b\t\tSet which cameras you want to trigger.example: [-b 0x0f-0xf0] the forward one is sync cameras which you want trigger,the after one is async cameras which you want trigger(all 8 cameras is setted sync model by default.like[-b 0xff-0])\n"
           "\t-p\t\tSet async cameras is triggered at which angle in a circle,not set by default.\n"
           "\t-h\t\tPrint this usage\n\n"
           "\tNOTE: It runs infinitely until you terminate it with <ctrl+c>\n");
}
/*demo程序命令相关参数设置接口，详细解释见README*/
static bool
parse_cmdline(int argc, char **argv) {
    int c;
    unsigned int tmp_w;
    unsigned int tmp_h;

    int sync_camera_bit_draw = 0,async_camera_bit_draw = 0;

    if (argc < 2) {
        print_usage();
        exit(EXIT_SUCCESS);
    }

    while ((c = getopt(argc, argv, "d:s:r:n:b:f:p:m:h")) != -1) {
        switch (c) {
            case 'd':
                strcpy(dev_node, optarg);
                break;
            case 's':
                if (sscanf(optarg, "%dx%d",
                           &tmp_w, &tmp_h) != 2) {
                    return false;
                }
                cam_w = tmp_w;
                cam_h = tmp_h;
                break;
                case 'f':
                camera_fmt_str = optarg;
                break;
            case 'm':
                camera_num = strtol(optarg, NULL, 10);
                break;
            case 'r':
                if (sscanf(optarg, "%hhu-%hhu",&stCameraCfgSend.sync_freq, &stCameraCfgSend.async_freq) != 2) {
                   print_usage();
                    return false;
                }
                printf("sync_freq : %d async_freq:%d\n",stCameraCfgSend.sync_freq,stCameraCfgSend.async_freq);
                break;
            case 'n':
                if (sscanf(optarg, "%hhu-%hhu",&stCameraCfgSend.sync_camera_num, &stCameraCfgSend.async_camera_num) != 2) {
                   print_usage();
                    return false;
                }
                           printf("sync_camera_num : %d async_camera_num:%d\n",(stCameraCfgSend.sync_camera_num),(stCameraCfgSend.async_camera_num));
                break;
            case 'b':
                if (sscanf(optarg, "%x-%x",&sync_camera_bit_draw, &async_camera_bit_draw) != 2) {
                   print_usage();
                   return false;
                }
                   stCameraCfgSend.sync_camera_bit_draw = (unsigned char)sync_camera_bit_draw;
                   stCameraCfgSend.async_camera_bit_draw = (unsigned char)async_camera_bit_draw;
                   printf("sync_camera_bit_draw : %d async_camera_bit_draw:%d\n",stCameraCfgSend.sync_camera_bit_draw,stCameraCfgSend.async_camera_bit_draw);
                break;
           case 'p':
               if (sscanf(optarg, "%hhu-%hhu-%hhu-%hhu-%hhu-%hhu-%hhu-%hhu",
                          &stCameraCfgSend.async_camera_pos[0], &stCameraCfgSend.async_camera_pos[1],&stCameraCfgSend.async_camera_pos[2],
                          &stCameraCfgSend.async_camera_pos[3],&stCameraCfgSend.async_camera_pos[4],&stCameraCfgSend.async_camera_pos[5]
                          ,&stCameraCfgSend.async_camera_pos[6],&stCameraCfgSend.async_camera_pos[7]) != 8) {
                   print_usage();
                   return false;
               }
                  printf("pos:[0]:%hhu [1]:%hhu [2]:%hhu [3]:%hhu [4]:%hhu [5]:%hhu [6]:%hhu [7]:%hhu \n",stCameraCfgSend.async_camera_pos[0],stCameraCfgSend.async_camera_pos[1],stCameraCfgSend.async_camera_pos[2],
                  stCameraCfgSend.async_camera_pos[3],stCameraCfgSend.async_camera_pos[4],stCameraCfgSend.async_camera_pos[5],stCameraCfgSend.async_camera_pos[6],stCameraCfgSend.async_camera_pos[7]);
               break;
            case 'h':
                print_usage();
                exit(EXIT_SUCCESS);
                break;
            default:
                print_usage();
                return false;
        }
    }
    return true;
}
/*demo程序主函数，分别打开n个窗口，并通过反复分别调用GetImageCvMat和GetImagePtr接口获取图像和时间戳在窗口中显示*/
int main(int argc, char *argv[]) {

    camera_context_t ctx[8] = {};

    stCameraCfgSend.async_camera_num = 0;
    stCameraCfgSend.async_freq = 0;
    stCameraCfgSend.async_camera_bit_draw = 0;
    stCameraCfgSend.sync_camera_num = 8;
    stCameraCfgSend.sync_freq = 30;
    stCameraCfgSend.sync_camera_bit_draw = 0xff;

    if (!parse_cmdline(argc, argv)) {
        return -1;
    }
    setenv("CHECK_TIME","0",0);
    char dev_node_tmp = dev_node[10];
    for(int i = 0; i < camera_num; i++){
        dev_node[10] = dev_node_tmp + i;
        ctx[i].dev_node = dev_node;
        ctx[i].camera_fmt_str = camera_fmt_str;
        ctx[i].output_fmt_str = output_fmt_str;
        ctx[i].cam_w = cam_w;
        ctx[i].cam_h = cam_h;
        ctx[i].out_w = out_w;
        ctx[i].out_h = out_h;
    }
    miivii::MvGmslCamera mvcam(ctx, camera_num, stCameraCfgSend);

    std::string windowName("DisplayCamera ");
    for (uint32_t i = 0; i < camera_num; i++) {
        cv::namedWindow(windowName + std::to_string(i), cv::WindowFlags::WINDOW_AUTOSIZE);
        cv::moveWindow(windowName + std::to_string(i), 200 * i, 200 * i);
    }
    cv::Mat outMat[camera_num];
    uint8_t *outbuf[camera_num];
    cv::Mat imgbuf[camera_num];
    signal(SIGINT, &handler);
    bool quit = false;
    uint64_t timestamp;
    while (!quit) {
        if (exitRequested) {
            quit = true;
            break;
        }
        uint8_t camera_no = dev_node[10] - 0x30;
        /*use cv data to get image*/
        if (mvcam.GetImageCvMat(outMat, timestamp, camera_no)) {
            for (uint32_t i = 0; i < camera_num; i++) {
                if ( ctx[i].output_fmt_str == "UYVY") {
                    cv::cvtColor(outMat[i], outMat[i], cv::COLOR_YUV2BGR_UYVY);
                    cv::imshow(windowName + std::to_string(i), outMat[i]);
                } else if (ctx[i].output_fmt_str == "ABGR32") {
                    cv::cvtColor(outMat[i], outMat[i], cv::COLOR_RGBA2BGR);
                    cv::imshow(windowName + std::to_string(i), outMat[i]);
                }
                CheckTimeStampLog(timestamp,dev_node_tmp - 0x30 + i);
            }
        } else {
            std::cerr << "Can't get image form camera." << std::endl;
        }
        if (cv::waitKey(1) == 27) {// Wait for 'esc' key press to exit
            break;
        }
        /*use raw data to get image*/
        if (mvcam.GetImagePtr(outbuf, timestamp, camera_no)) {
            for (uint32_t i = 0; i < camera_num; i++) {
                if ( ctx[i].output_fmt_str == "UYVY") {
                    imgbuf[i] = cv::Mat(out_h, out_w, CV_8UC2, outbuf[i]);
                    cv::Mat mrgba(out_h, out_w, CV_8UC3);
                    cv::cvtColor(imgbuf[i], mrgba, cv::COLOR_YUV2BGR_UYVY);
                    cv::imshow(windowName + std::to_string(i), mrgba);
                } else if (ctx[i].output_fmt_str == "ABGR32") {
                    imgbuf[i] = cv::Mat(out_h, out_w , CV_8UC4, outbuf[i]);
                    cv::cvtColor(imgbuf[i], imgbuf[i], cv::COLOR_RGBA2BGR);
                    cv::imshow(windowName + std::to_string(i), imgbuf[i]);
                }
                CheckTimeStampLog(timestamp,dev_node_tmp - 0x30 + i);
            }
        } else {
            std::cerr << "Can't get image form camera." << std::endl;
        }
        if (cv::waitKey(1) == 27) {// Wait for 'esc' key press to exit
            break;
        }
    }
    return 0;
}
